# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/thebes/gfxPlatformFontList.h
# Commit: 0d746825f92a
# Full Hash: 0d746825f92a3039433aa275997c6e7f2ebfaf56
# Author: Jonathan Kew <jkew@mozilla.com>
# Date: 2021-05-18 21:36:28
# Regressor Bug: 1708768
# File Overlap Count: 2
# Description:
#   Bug 1708768 - Make InitFontList safe to call off-main-thread on macOS, and run font-list initialization on a separate thread during startup. r=lsalzman
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D115295
# ==============================================================================

diff -r fbe7310ee5f3 -r 0d746825f92a gfx/thebes/gfxPlatformFontList.h
--- a/gfx/thebes/gfxPlatformFontList.h	Tue May 18 11:10:43 2021 +0000
+++ b/gfx/thebes/gfxPlatformFontList.h	Tue May 18 11:10:43 2021 +0000
@@ -170,30 +170,36 @@
   typedef nsTArray<FontFamily> PrefFontList;
 
   static gfxPlatformFontList* PlatformFontList() {
-    // Currently, only macOS uses lazy font-list initialization; on other
-    // platforms we create it during gfxPlatform::Init().
-    if (!sPlatformFontList) {
-      if (!gfxPlatform::GetPlatform()->CreatePlatformFontList()) {
+    if (sPlatformFontList->IsInitialized()) {
+      return sPlatformFontList;
+    }
+    // Currently, only macOS uses OMT font-list initialization; on other
+    // platforms we initialize it directly during gfxPlatform::Init().
+    if (sInitFontListThread) {
+      PR_JoinThread(sInitFontListThread);
+      sInitFontListThread = nullptr;
+      // If font-list initialization failed, the thread will have cleared
+      // the static sPlatformFontList pointer; we cannot proceed without any
+      // usable fonts.
+      if (!sPlatformFontList) {
         MOZ_CRASH("Could not initialize gfxPlatformFontList");
       }
     }
+    if (!sPlatformFontList->InitFontList()) {
+      MOZ_CRASH("Could not initialize gfxPlatformFontList");
+    }
     return sPlatformFontList;
   }
 
-  static bool Initialize(gfxPlatformFontList* aList) {
-    sPlatformFontList = aList;
-    if (aList->InitFontList()) {
-      return true;
-    }
-    Shutdown();
-    return false;
-  }
+  static bool Initialize(gfxPlatformFontList* aList);
 
   static void Shutdown() {
     delete sPlatformFontList;
     sPlatformFontList = nullptr;
   }
 
+  bool IsInitialized() const { return mFontlistInitCount; }
+
   virtual ~gfxPlatformFontList();
 
   // Initialize font lists; return true on success, false if something fails.
@@ -523,6 +529,11 @@
   // (The legacy (non-shared) list just returns 0 here.)
   uint32_t GetGeneration() const;
 
+  // Sometimes we need to know if we're on the InitFontList startup thread.
+  static bool IsInitFontListThread() {
+    return PR_GetCurrentThread() == sInitFontListThread;
+  }
+
  protected:
   friend class mozilla::fontlist::FontList;
   friend class InitOtherFamilyNamesForStylo;
@@ -719,9 +730,6 @@
                                            StretchRange aStretchForEntry,
                                            SlantStyleRange aStyleForEntry);
 
-  // commonly used fonts for which the name table should be loaded at startup
-  virtual void PreloadNamesList();
-
   // load the bad underline blocklist from pref.
   void LoadBadUnderlineList();
 
@@ -820,13 +828,13 @@
   FontFamilyTable mOtherFamilyNames;
 
   // flag set after InitOtherFamilyNames is called upon first name lookup miss
-  bool mOtherFamilyNamesInitialized;
+  bool mOtherFamilyNamesInitialized = false;
 
   // The pending InitOtherFamilyNames() task.
   RefPtr<mozilla::CancelableRunnable> mPendingOtherFamilyNameTask;
 
   // flag set after fullname and Postcript name lists are populated
-  bool mFaceNameListsInitialized;
+  bool mFaceNameListsInitialized = false;
 
   struct ExtraNames {
     ExtraNames() = default;
@@ -873,20 +881,22 @@
 
   // data used as part of the font cmap loading process
   nsTArray<RefPtr<gfxFontFamily>> mFontFamiliesToLoad;
-  uint32_t mStartIndex;
-  uint32_t mNumFamilies;
+  uint32_t mStartIndex = 0;
+  uint32_t mNumFamilies = 0;
 
   // xxx - info for diagnosing no default font aborts
   // see bugs 636957, 1070983, 1189129
-  uint32_t mFontlistInitCount;  // num times InitFontList called
+  uint32_t mFontlistInitCount = 0;  // num times InitFontList called
 
   nsTHashSet<gfxUserFontSet*> mUserFontSetList;
 
-  nsLanguageAtomService* mLangService;
+  nsLanguageAtomService* mLangService = nullptr;
 
   nsTArray<uint32_t> mCJKPrefLangs;
   nsTArray<mozilla::StyleGenericFontFamily> mDefaultGenericsLangGroup;
 
+  nsTArray<nsCString> mEnabledFontsList;
+
   mozilla::UniquePtr<mozilla::fontlist::FontList> mSharedFontList;
 
   nsClassHashtable<nsCStringHashKey, mozilla::fontlist::AliasData> mAliasTable;
@@ -903,7 +913,9 @@
 
   FontVisibility mVisibilityLevel = FontVisibility::Unknown;
 
-  bool mFontFamilyWhitelistActive;
+  bool mFontFamilyWhitelistActive = false;
+
+  static PRThread* sInitFontListThread;
 };
 
 MOZ_MAKE_ENUM_CLASS_BITWISE_OPERATORS(gfxPlatformFontList::FindFamiliesFlags)